### PR TITLE
fix(evm): EIP-7002 withdrawals decoding

### PR DESCRIPTION
**Pull request by shekhirin**
This PR fixes the [EIP-7002 system contract call output decoding](https://eips.ethereum.org/EIPS/eip-7002#execution-layer), and also adds a test that queues a deposit and then asserts the requests in the executor output.

### CODE REVIEW

File: crates/revm/src/state_change.rs [hunk 1]
```diff
@@ -329,9 +330,25 @@ where
 
     let withdrawal_requests = match result {
         ExecutionResult::Success { output, .. } => {
-            Vec::<Request>::decode(&mut output.into_data().as_ref()).map_err(|err| {
-                BlockValidationError::WithdrawalRequestsContractCall { message: err.to_string() }
-            })
+            let mut requests = vec![];
```
Comment (mattsse) [inline crates/revm/src/state_change.rs:333]: can we document the output format here

Comment (mattsse) [inline crates/revm/src/state_change.rs:333]: can we use with_capacity data.len() / 20 + 8 + 48 ?

**Review by mattsse — CHANGES_REQUESTED** [review]
nits

File: crates/ethereum/evm/src/execute.rs [hunk 2]
```diff
@@ -841,4 +869,78 @@ mod tests {
...
+                .shanghai_activated()
+                .with_fork(Hardfork::Prague, ForkCondition::Timestamp(0))
+                .build(),
+        );
+
+        let mut db = create_state_provider_with_withdrawal_requests_contract();
+
+        let secp = Secp256k1::new();
+        let sender_key_pair = Keypair::new(&secp, &mut generators::rng());
+        let sender_address = public_key_to_address(sender_key_pair.public_key());
```
Comment (mattsse) [inline crates/ethereum/evm/src/execute.rs:886]: why is this needed?

to generate a random address?

can we use Address::random?

File: crates/revm/src/batch.rs [hunk 3]
```diff
@@ -83,6 +83,11 @@ impl BlockBatchRecord {
         std::mem::take(&mut self.receipts)
     }
 
+    /// Returns the recorded requests.
+    pub fn requests(&self) -> &Vec<Requests> {
```
Comment (mattsse) [inline crates/revm/src/batch.rs:87]:
```suggestion
    pub fn requests(&self) -> &[Requests] {
```

File: crates/revm/src/state_change.rs [hunk 4]
```diff
@@ -329,9 +330,32 @@ where
...
+            let mut requests = Vec::with_capacity(data.len() / (20 + 48 + 8));
+
+            // Withdrawals are encoded as a series of withdrawal requests, each with the following
+            // format:
+            //
+            // +------+--------+--------+
+            // | addr | pubkey | amount |
+            // +------+--------+--------+
+            //    20      48        8
+            while data.has_remaining() {
```
Comment (mattsse) [inline crates/revm/src/state_change.rs]: this loop will panic if data.len % 20 + 48 + 8 != 0

I assume this is not possible, but maybe not worth checking?

**Review by mattsse — APPROVED** [review]

File: crates/ethereum/evm/src/execute.rs (hunk 2 above)
Comment (shekhirin) [reply crates/ethereum/evm/src/execute.rs:886]: i'm signing the transaction with it

File: crates/revm/src/state_change.rs [hunk 5]
```diff
@@ -318,21 +319,22 @@ where
         Err(e) => {
             evm.context.evm.env = previous_env;
             return Err(BlockValidationError::WithdrawalRequestsContractCall {
-                message: e.to_string(),
+                message: format!("execution failed: {e}"),
```
Comment (gakonst) [inline crates/revm/src/state_change.rs:322]: `execution failed` should be part of the error

Comment (shekhirin) [reply crates/revm/src/state_change.rs:322]: not really, we don't have a transaction hash to attach, and it's an irregular state transition, i.e. not a real transaction https://github.com/paradigmxyz/reth/blob/308f979402e638c6ac5494d38230461a2f4fc0d3/crates/interfaces/src/executor.rs#L11-L19
